src/invoice_utils/web.py

```diff
@@ -83,20 +89,24 @@ def generate_invoice(request: InvoiceRequest):
     root_dir = Path(__file__).parent
     basic_rules = str(root_dir / "basic.json")
     engine = InvoicingEngine(basic_rules)
-    _send_mail(request)
-    return engine.process(
+    context = engine.process(
         int(request.header.number), request.header.timestamp, request.items
     )
+    renderer = PdfInvoiceRenderer("invoice")
+    invoice_path = root_dir / f"{request.header.timestamp:%Y%m%d}-{int(request.header.number):04}-invoice.pdf"
+    renderer.render(context, str(invoice_path))
```

Comment:
What happens to the file that was generated after the HTTP response was sent to the client? I would expect it's being removed. A `ContextManager` would be nice here.

Reply:
Another topic relates to having write access on disk. Please test that not having write access on disk doesn't crash the web endpoint. The solution is completely up to you and can range from using streams instead of files to simply logging a warning related to not having write access.
The response differs depending on the chosen solution. If you want to keep using files, the server should return 503 or 507 depending on the reason why it can't write the file to disk.
Of course, the tests should be written first.

Reply:
For this endpoint, the invoice creation on disk is not required so I updated the render method with a flag 'persist'. I created the functionality to check the write access of a configurable location to persist on disk, which may be used in the future. Perhaps this endpoint is doing too much at the moment and in the future could be split into an invoice generation endpoint and a send_invoice\invoice_id endpoint.